core-api/src/lib.rs

```diff
@@ -45,13 +45,17 @@ pub trait Worker: Send + Sync {
     /// concurrently internally.
     async fn poll_activity_task(&self) -> Result<ActivityTask, PollActivityError>;
 
-    /// Tell the worker that a workflow activation has completed. May be freely called concurrently.
+    /// Tell the worker that a workflow activation has completed. May (and should) be freely called
+    /// concurrently. The future may take some time to resolve, as fetching more events might be
+    /// necessary for completion to... complete - thus SDK implementers should make sure they do
+    /// not serialize completions.
     async fn complete_workflow_activation(
```

Comment:
Strictly speaking, it's not necessary - but it also has no meaningful downside. Either way things can't proceed until that fetching happens, and since callers are expected to not serialize completion calls, all it does is add another task and a bit of complexity to the internals without any meaningful benefit.

Reply:
If you think it adds needless complexity, I'm fine with this change.
Just weird to make to tie the 2 operations IMHO, I personally wouldn't.